### PR TITLE
Add `drop_empty_moments` and `drop_negligible_operations` transformers.

### PR DESCRIPTION
**Pull request by tanujkhattar**
- Replaces `DropEmptyMoments` and `DropNegligible` optimizers.
- Part of Organization (and deprecation) of cirq-core/optimizers in cirq-core/transformers #4722
- Follows the new Transformer API Compiling: Circuit Transformers API #4483
- Supports no compile tags NoCompile Tag for optimizers NoCompile Tag for optimizers #4253

### CODE REVIEW

File: cirq-core/cirq/optimizers/merge_interactions_test.py [hunk 1]
```diff
@@ -30,13 +30,19 @@ def assert_optimizes(before: cirq.Circuit, expected: cirq.Circuit):
...
         post(actual)
         post(expected)
 
+    followup_optimizations_new: List[cirq.TRANSFORMER] = [
+        cirq.drop_negligible_operations,
+        cirq.drop_empty_moments,
+    ]
+    for post_new in followup_optimizations_new:
+        actual = post_new(actual).unfreeze(copy=False)
+        expected = post_new(expected).unfreeze(copy=False)
```
Comment (maffoo) [inline cirq-core/cirq/optimizers/merge_interactions_test.py:44]: I'd suggest using the "transformer" terminology here:
```suggestion
    followup_transformers: List[cirq.TRANSFORMER] = [
        cirq.drop_negligible_operations,
        cirq.drop_empty_moments,
    ]
    for transform in followup_transformers:
        actual = transform(actual).unfreeze(copy=False)
        expected = transform(expected).unfreeze(copy=False)
```

File: cirq-core/cirq/optimizers/merge_interactions_to_sqrt_iswap_test.py [hunk 2]
```diff
@@ -41,13 +41,19 @@ def assert_optimizes(before: cirq.Circuit, expected: cirq.Circuit, **kwargs):
...
         post(actual)
         post(expected)
 
+    followup_optimizations_new: List[cirq.TRANSFORMER] = [
+        cirq.drop_negligible_operations,
+        cirq.drop_empty_moments,
+    ]
+    for post_new in followup_optimizations_new:
+        actual = post_new(actual).unfreeze(copy=False)
+        expected = post_new(expected).unfreeze(copy=False)
```
Comment (maffoo) [inline cirq-core/cirq/optimizers/merge_interactions_to_sqrt_iswap_test.py:55]: Here, too, use "transformer" for variable names.

File: cirq-core/cirq/optimizers/merge_single_qubit_gates_test.py [hunk 3]
```diff
@@ -29,10 +29,10 @@ def assert_optimizes(
     optimizer(before)
 
     # Ignore differences that would be caught by follow-up optimizations.
-    followup_optimizations = [cirq.DropNegligible(), cirq.DropEmptyMoments()]
+    followup_optimizations = [cirq.drop_negligible_operations, cirq.drop_empty_moments]
     for post in followup_optimizations:
```
Comment (maffoo) [inline cirq-core/cirq/optimizers/merge_single_qubit_gates_test.py:33]:
```suggestion
    followup_transformers = [cirq.drop_negligible_operations, cirq.drop_empty_moments]
    for transform in followup_transformers:
```

File: cirq-core/cirq/transformers/drop_empty_moments_test.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                    cirq.Moment([cirq.CNOT(q1, q2)]),
+                    cirq.Moment(),
+                ]
+            )
+        ),
+        cirq.Circuit(
+            [
+                cirq.Moment([cirq.CNOT(q1, q2)]),
+            ]
+        ),
```
Comment (maffoo) [inline cirq-core/cirq/transformers/drop_empty_moments_test.py:36]: nit: can use multi-arg `Circuit` and `Moment` constructors to reduce indentation:
```suggestion
        cirq.drop_empty_moments(
            cirq.Circuit(
                cirq.Moment(),
                cirq.Moment(),
                cirq.Moment(cirq.CNOT(q1, q2)),
                cirq.Moment(),
            )
        ),
        cirq.Circuit(cirq.Moment(cirq.CNOT(q1, q2))),
```

File: cirq-core/cirq/transformers/drop_negligible_operations_test.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+# limitations under the License.
+
+import cirq
+
+NO_COMPILE_TAG = "no_compile_tag"
+
+
+def test_leaves_big():
+    a = cirq.NamedQubit('a')
+    circuit = cirq.Circuit([cirq.Moment([cirq.Z(a) ** 0.1])])
```
Comment (maffoo) [inline cirq-core/cirq/transformers/drop_negligible_operations_test.py:22]: nit: can remove lists
```suggestion
    circuit = cirq.Circuit(cirq.Moment(cirq.Z(a) ** 0.1))
```
Same thing applies throughout this module.

**Review by maffoo — APPROVED** [review]
Minor comments, then LGTM.

**Comment by tanujkhattar** [comment]
@maffoo Thanks for the review! Addressed all comments, merging now.

**Pull request merged**